### ISSUE TITLE
Save results for different hyperparameters, etc. to different places to avoid overwriting older results

### ISSUE THREAD

**Issue by AndyMcAliley** [USGS-R/lake-temperature-lstm-static#18]
Following up on [this discussion](https://github.com/USGS-R/lake-temperature-lstm-static/discussions/14#discussioncomment-2424230), we want to save multiple sets of results that correspond to different choices of hyperparameters, processing decisions, etc. This issue is also linked to the question of how to log and track those choices alongside results.

**Comment by AndyMcAliley** [comment]
Here's one approach: we could create a table of inputs - a csv with one row per trained model. A pre-trained model and a fine-tuned model could have separate rows in this table. The columns of the table list every pertinent choice or variable: which data source was used to form training data, choices of how to process the training data, model hyperparameters, git commit hash, whether the git working directory was clean, etc. Let's call these variables "settings". Each row would also have a unique index. Our pipeline could read this table in as an input and define targets based on row index. 

So, say I wanted to train a model with new settings. First, I'd add a new row to the table of inputs, specifying every setting. Say that went into row 3. To train the model with the new settings, I'd run
```
snakemake --cores all metrics_3.csv
```

**Comment by lindsayplatt** [comment]
@hcorson-dosch did something similar for the recent targets pipeline to use the 6 GCMs. I wonder if she can comment here about how it worked and what some of the challenges might be.

**Comment by hcorson-dosch** [comment]
Ok - as noted earlier, our approach was very much defined by `targets`, but I'll share a bit about it. It's been working well for us.

The idea was to provide the model execution function with a model configuration dataframe with one row for each run to (1) document the inputs to each model run, then, within the model execution function, add to that table to (2) Track details about the run itself and whether it succeeded, and (3) Track the output of the model run and info needed to package up that output.

The modified dataframe was returned by the model execution function, which also saved model output. Returning a dataframe meant we could keep the target as an object target (retaining all this key info and making use of it downstream), while still tracking the output files `export_fl` and the data they contain `export_fl_hash` (for which we'd normally use a file target). It's a bit of a hacky `targets` approach that Lindsay and I came up with, but so far, so good!

Going in to the model run, the model configuration dataframe has the following (which of the three purposes each item met noted in brackets):
* [1, 2, 3] **site_id**
* [1] **state**
* [1, 2, 3] **gcm** -- name of the gcm    
* [1, 2, 3] **time_period** -- the 20yr time period
* [3] **gcm_start_date** -- the start date of the raw driver data (without burn-in)
* [3] **gcm_end_date** -- the end date of the raw driver data (without burn-out)
* [2] **burn_in**  - length of the burn in period (in days)
* [1, 2] **burn_in_start** - date the the simulation should start (gcm_start_date -  burn-in)
* [2] **burn_out** - length of burn out (in days)
* [1, 2] **burn_out_end** - date the the simulation should end (gcm_end_date + burn-out)
* [1, 2] **raw_meteo_fl** -- the filepath to the driver data (with burn in/out)
* [1] meteo_fl_hash -- hash of the meteo data (if it changes, will trigger re-run)

And here's what's in the dataframe returned by the model execution function (which of the three purposes each item met noted in brackets):
- [1, 2, 3] **site_id**
- [1, 2, 3] **gcm** -- name of the gcm
- [1, 2, 3] **time_period** -- the 20yr time period
- [1, 2] **raw_meteo_fl** -- the filepath to the driver data (with burn in/out)
- [3] **gcm_start_date** -- the start date of the raw driver data (without burn-in)
- [3] **gcm_end_date** -- the end date of the raw driver data (without burn-out)
- [2] **burn_in**  - length of the burn in (in days)
- [2] **burn_out** - length of burn out (in days)
- [1, 3] **export_fl** - filepath of the output file from the model run
- [3] **export_fl_hash** - hash of the output file
- [2] **extraction_error** - tracking whether or not there was an error extracting the model output
- [2] **glm_run_date** - date and time of model run
- [2] **glm_version** - version of GLM used
- [2] **glm_time_s**- length of time for GLM to run
- [2] **param_sim_start** - date of simulation start (gcm_start_date -  burn-in)
- [2] **param_sim_stop** - date of simulation end (gcm_end_date + burn out)
- [2] **max_output_date** - max date of model output (sometimes doesn't match)
- [2] **glm_code** - code returned by model
- [2, 3] **glm_success** - T/F did the model run succeed

On HPC I also have a few additional columns that track how many seconds various pieces of the run took to complete.

So essentially the returned dataframe tracks all this info about the model run (Purpose 2)

Purpose 3 came into play we then make a new datfarame where we select only the variables from the full returned dataframe that relate to model output [3], group the dataframe by `site_id` and `gcm`, filter to identify only fully successful runs, and then use the resulting dataframe to determine what output to package up.  The idea was that the `export_file_hash` column is the only of the saved variables that would change if the actual model output differed, so if something changed in the code, we'd only rebuild the output if the actual data stored in the export file changed. That's where I think this approach deviates pretty far from what you'd implement in `Snakemake`, given that Snakemake a) doesn't make use of object targets and b) tracks file timestamps rather than content.

Anyhow, not sure how helpful that is, but I'm happy to go into more detail or chat about it.

**Comment by AndyMcAliley** [comment]
Thanks Hayley! Delineating the 3 purposes especially helps, and that maps well onto this project.

**Comment by AndyMcAliley** [comment]
Like Hayley said, one issue is that `Snakemake` works mainly with input and output files, not objects. A way around that would be to create a separate input file for every new run, instead of adding rows to a single tabular file. We could take a similar approach to outputs and create a new output file that essentially consists of one row of the output data frame that Hayley described above.

Actually, that addresses both of the differences between `Snakemake` and `targets` that Hayley brought up:
a) We use files as inputs instead of objects.
b) Tracking file timestamps will work for this approach.

PROs
- Simple pipeline: one input settings file and one output settings file per trained model
- Easy to create summary tables of all runs so far
- Results don't get overwritten

CONs
- would recreate training data even when it's not needed
- need to rewrite some Snakemake code to use input settings file in place of config
- requires and generates a lot of small files
- doesn't take advantage of `Snakemake`'s `expand()` function, which could do grid searches more elegantly.

**Comment by AndyMcAliley** [comment]
# "Config" approach

Another approach is to specify settings through the `Snakemake` .yaml configuration files. This is currently how the pipeline is set up. Each time the pipeline is run, we'd assign a run number to the results (either automatically or manually). The trained model(s) and all of the settings (input and output) would be saved to a new directory named by run number. The inputs settings would be read in from the relevant .yaml configuration files. 

`Snakemake`'s `expand()` function can execute a grid search in parallel, resulting in many trained models (Jeff's blog post has a nice example; draft version [here](https://github.com/jsadler2/wdfn-blog/blob/snakemake-dl-experiments/content/snakemake-ml-experiments.md)). In the "config" approach, I think it would be easier to consider the whole grid search and all the resulting models as one run. So, one run could result in one or many trained models. 

To ease reproducibility, we can also copy the .yaml configuration files into the new run directory. 

## PROs
- Could use `Snakemake`'s `expand()` to execute grid searches in parallel
- Snakefiles are already written to use config files - no rewrite needed
- Training data only gets remade when necessary
- More closely mirrors Andy's mental model (YMMV!)

## CONs
- Could become more ad hoc: some "runs" may have multiple models, and naming conventions within runs could get tricky if there are multiple models
- Hard to parallelize across arbitrary changes in settings (though easier to parallelize over grid searches)
- Potentially easier to accidentally overwrite results (depending on implementation)
- To reproduce an earlier run, config files must be manually changed or copied into the right place.

**Comment by AndyMcAliley** [comment]
Okay, the more I think about this, the clearer I get. We can separate out the input settings file structure (Table vs Split vs Config) from the use of `expand()` and the output file structure (one run per model vs one run per experiment, where an experiment could result in a single model or many models). In truth, any input settings format could support any output format. So, break it down like this:
1. Input settings format
    a. "Table"
    b. "Split"
    c. "Config"
2. Output format
    a. One run per model
    b. One run per experiment

This view softens some of the PROs and CONs above. For instance, the Config approach would accommodate `expand()` more readily than the "Split" approach, but `expand()` could be used with either.

**Comment by AndyMcAliley** [comment]
Lindsay, Hayley, Jordan and I met this afternoon to discuss this issue. In the spirit of going forward with a solution that is easy to implement, flexible, and not overly-polished, we decided the following:
1. Use the "Config" input approach
2. Use the "One run per experiment" output approach

For each experiment, we can tag its corresponding git commit. Run numbers (used to determine output folder) can be specified in the configuration file. If a grid search is performed, we don't have to keep every trained model - we could just go back and delete the files containing weights and biases manually, keeping only the best performing model. But, we should keep the output csv containing input settings, performance metrics, etc. for every model we train. It would be useful to write code that concatenates all the output csvs for all the runs into one summary csv.
